File: VBNetCore.Mvvm/ViewModels/HasPopupValidationViewModel.vb
Imports VBNetCore.Mvvm.Abstractions
Imports System
Imports System.Threading
Imports System.Windows.Threading

Namespace VBNetCore.Mvvm.ViewModels

    Public MustInherit Class HasPopupValidationViewModel
        Inherits ValidationPropertyChangedBase
        Implements IPopupable
        Public Property CurrentPopup As PopupViewModel Implements IPopupable.CurrentPopup
            Get
                Return GetProperty(Of PopupViewModel)()
            End Get
            Set(ByVal value As PopupViewModel)
                SetProperty(value)
            End Set
        End Property
        Public Property IsPopupVisible As Boolean Implements IPopupable.IsPopupVisible
            Get
                Return GetProperty(Of Boolean)()
            End Get
            Set(ByVal value As Boolean)
                SetProperty(value)
            End Set
        End Property
        Public Sub HidePopup() Implements IPopupable.HidePopup
            IsPopupVisible = False
            If CurrentPopup IsNot Nothing Then
                RemoveHandler CurrentPopup.CloseHandler, AddressOf PopupViewModel_CloseHandler
                CurrentPopup = Nothing
            End If
        End Sub

        Public Function ShowPoup(ByVal popupViewModel As PopupViewModel) As PopupViewModel Implements IPopupable.ShowPoup
            AddHandler popupViewModel.CloseHandler, AddressOf PopupViewModel_CloseHandler
            CurrentPopup = popupViewModel
            IsPopupVisible = True

            While IsPopupVisible
                If Dispatcher.CurrentDispatcher.HasShutdownStarted OrElse Dispatcher.CurrentDispatcher.HasShutdownFinished Then Exit While
                Call Dispatcher.CurrentDispatcher.Invoke(DispatcherPriority.Background, New ThreadStart(Sub()
                                                                                                        End Sub))
                Thread.Sleep(20)
            End While

            Return popupViewModel
        End Function
        Private Sub PopupViewModel_CloseHandler(ByVal sender As Object, ByVal e As EventArgs)
            HidePopup()
        End Sub
    End Class
End Namespace

File: VBNetCore.Mvvm/ViewModels/HasPopupValidationViewModel.cs
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Controls;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Windows.Threading;

namespace NetCore.Mvvm.ViewModels
{

    public abstract class HasPopupValidationViewModel : ValidationPropertyChangedBase, IPopupable
    {
        public PopupViewModel? CurrentPopup 
        { 
            get => GetProperty<PopupViewModel>();
            set => SetProperty(value);
        }
        public bool IsPopupVisible 
        { 
            get => GetProperty<bool>();
            set => SetProperty(value);
        }
        public void HidePopup()
        {
            IsPopupVisible = false;
            if (CurrentPopup != null)
            {
                CurrentPopup.CloseHandler -= PopupViewModel_CloseHandler;
                CurrentPopup = null;
            }
        }

        public PopupViewModel? ShowPoup(PopupViewModel popupViewModel)
        {
            popupViewModel.CloseHandler += PopupViewModel_CloseHandler;
            CurrentPopup = popupViewModel;
            IsPopupVisible = true;

            while (IsPopupVisible)
            {
                if (Dispatcher.CurrentDispatcher.HasShutdownStarted || Dispatcher.CurrentDispatcher.HasShutdownFinished) break;
                Dispatcher.CurrentDispatcher.Invoke(DispatcherPriority.Background, new ThreadStart(delegate { }));
                Thread.Sleep(20);
            }

            return popupViewModel;
        }
        private void PopupViewModel_CloseHandler(object? sender, EventArgs e)
        {
            HidePopup();
        }
    }
}
